lib/Monads/trace/Trace_RG.thy

```diff
@@ -230,6 +314,135 @@ lemma bind_twp[wp_split]:
   apply (clarsimp simp: guar_cond_append)
   done
 
+lemma bindE_twp[wp_split]:
+  "\<lbrakk>\<And>r. \<lbrace>Q' r\<rbrace>,\<lbrace>R\<rbrace> g r \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>; \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q'\<rbrace>,\<lbrace>E\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f >>=E (\<lambda>r. g r) \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  apply (simp add: bindE_def validIE_def)
+  apply (erule bind_twp[rotated])
+  apply (clarsimp simp: lift_def throwError_def split: sum.splits)
+  apply (subst validI_valid_no_trace_eq)
+   apply wpsimp+
+  done
+
+\<comment> \<open>FIXME MC: improve naming for bind_twp vs seq_ext vs rg_seq_ext
+          existing proofs mention hoare_seq_ext a lot and seq_ext a few times\<close>
+\<comment> \<open>FIXME MC: names of seq_ext and seq_extE clash with valid equivalents\<close>
+lemmas seq_ext = bind_twp[rotated]
+lemmas seq_extE = bindE_twp[rotated]
+
+lemmas rg_seq_ext = bind_twp
+lemmas rg_seq_extE = bindE_twp
+
+lemma rg_TrueI[simp]:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f -,\<lbrace>\<lambda>_ _. \<top>\<rbrace> = prefix_closed f"
+  by (simp add: validI_def)
+
+lemma rgE_TrueI[simp]:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f -,\<lbrace>\<lambda>_ _. \<top>\<rbrace>,\<lbrace>\<lambda>_ _. \<top>\<rbrace> = prefix_closed f"
```

Comment:
These two look slightly dangerous as `[simp]`, because they could leave `P` and `R` as free schematics in the goal.

Reply:
All of these `validI` rules given as `simp` are based on ones from nondet, in this case `hoare_TrueI`. In practice here, any proof with `P` as a schematic will almost definitely be hit with `wp` first, which will use `wp_post_tautI` to instantiate the `P` as `λ_ _. True`.

As you mentioned later on, `R` should almost never be schematic. At a guess we'll only see schematic relies in proofs dealing with `parallel` and in some refinement proofs, and in both of these cases the `R` would be instantiated pretty quickly.

Reply:
Having said that, should I take the opportunity to try and clean up some of these `[simp]` rules? That would be another thing to try in `nondet`, to see if they're actually used anywhere.

Reply:
`nondet` has a bunch of these from very early times, but removing then will cause breakage in strange places. I'd leave that for a little later and focus on the trace monad only for now. Maybe add a cleanup issue for it, but I'd try leaving them out of `[simp]` here so that we're at least not propagating them to the new monad.